### PR TITLE
Feature/optimize

### PR DESCRIPTION
**Pull request by purge**
- Optimized build.
- Accepts AMD or promise based module loader.
- Created nested structure from config options.
- Update core dependencies.
- Everything loaded in its own context
- Support pre-loading of plugins as mixins to main bundles.
- Tests

### CODE REVIEW

File: app/main.js [hunk 1]
```diff
@@ -1,100 +1,114 @@
...
+      if(bundle.Controller) {
+        var widget = new bundle.Controller({
+          sqwidget: this,
+          config: pkg
+        });
+
+        //bus events
+        bean.fire(this, "rendered:" + pkg.location);
+        bean.fire(this, "rendered:" + pkg.id);
+        bean.fire(pkg, "rendered");
```
Comment (adhipg) [inline app/main.js]: Namespacing using dots?
And in that case, do we need the last one?

File: package.json [hunk 2]
```diff
@@ -36,10 +37,14 @@
     "karma-mocha": "~0.1.0",
     "karma-sinon": "0.0.1",
     "karma-chai-plugins": "~0.1.3",
+    "grunt-contrib-uglify": "~0.2.5",
+    "grunt-contrib-concat": "~0.3.0",
+    "grunt-init": "# This is because of the yes option. PR raised",
```
Comment (adhipg) [inline package.json]: So are we doing the comments in the json thing?

File: Gruntfile.js [hunk 3]
```diff
@@ -24,24 +37,60 @@ module.exports = function(grunt) {
...
+          },
+          include: ['requirejs', 'sqwidget'],
+          // Wrapper for AMD
+          wrap: {
+            startFile: 'src/_wrapper/top.js',
+            endFile: 'src/_wrapper/bottom.js'
+          },
+          //optimize: 'uglify2',
+          optimize: 'none',
+          preserveLicenseComments: false,
```
Comment (adhipg) [inline Gruntfile.js:59]: Any particular reason for this?

File: Gruntfile.js [hunk 4]
```diff
@@ -24,24 +37,60 @@ module.exports = function(grunt) {
...
+      compile: {
+        options: {
+          baseUrl: "src",
+          out: 'dist/<%= bower.name %>.js',
+          paths: {
+            almond: 'lib/almond/almond',
+            requirejs: 'lib/requirejs/require',
+            domReady: 'lib/requirejs-domready/domReady',
+          },
+          include: ['requirejs', 'sqwidget'],
```
Comment (adhipg) [inline Gruntfile.js:50]: Why does the `requirejs` compile need `requirejs` again?

Comment (purge) [inline Gruntfile.js:50]: because its bundled as part of the binary, so it can be namespaced and not need data-main.

File: Gruntfile.js (hunk 3 above)
Comment (purge) [inline Gruntfile.js:59]: smaller.

Comment (adhipg) [inline Gruntfile.js:59]: But it makes using the libraries illegal with certain licenses.

File: Gruntfile.js (hunk 4 above)
Comment (adhipg) [inline Gruntfile.js:50]: right! Could there be a comment there because it does look out-of-place and it's actually a feature?

**Comment by adhipg** [comment]
This needs `development` merging back in for some reason.

File: grunt-scaffold/root/src/main.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
+define([
+  'require',
+  'i18n!./nls/strings',
+  'css!compiled/css/main.css',
+  'rv!templates/index.html'
+], function(require, strings, css, template ) {
+  return {
+    Controller: function(opts) {
+      this.view = new Ractive({
```
Comment (adhipg) [inline grunt-scaffold/root/src/main.js]: Is Ractive is now a dependency of sqwidget as a whole?

File: src/core.js [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+    if(bundle.Controller) {
+      var widget = new bundle.Controller({
+        sqwidget: this,
+        config: pkg
+      });
+
+      //bus events
+      bean.fire(this, "rendered:" + pkg.url, [bundle]);
+      bean.fire(this, "rendered:" + pkg.id,[bundle]);
```
Comment (adhipg) [inline src/core.js]: Isn't using dots better?

File: grunt-scaffold/root/src/main.js (hunk 5 above)
Comment (purge) [inline grunt-scaffold/root/src/main.js]: No this is just the scaffold

—

File: src/core.js (hunk 6 above)
Comment (purge) [inline src/core.js]: hmm, did we get consensus on event layout?

**Comment by adhipg** [comment]
The README.md file is outdated and obsolete. Can we please have one that corresponds to this version?

**Comment by adhipg** [comment]
Warnings about missing `grunt-shell`

**Comment by adhipg** [comment]
Please remove the default task in Gruntfile.js. It's just confusing and obsolete.

File: src/core.js (hunk 6 above)
Comment (adhipg) [inline src/core.js]: I think we did since bean namespaces by dot.

**Pull request merged**